flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismAndInputInfosDecider.java

```diff
@@ -0,0 +1,479 @@
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Default implementation of {@link VertexParallelismAndInputInfosDecider}. This implementation will
+ * decide parallelism and {@link JobVertexInputInfo}s as follows:
+ *
+ * <p>1. For job vertices whose inputs are all ALL_TO_ALL edges, evenly distribute data to
+ * downstream subtasks, make different downstream subtasks consume roughly the same amount of data.
+ *
+ * <p>2. For other cases, evenly distribute subpartitions to downstream subtasks, make different
+ * downstream subtasks consume roughly the same number of subpartitions.
+ */
+public class DefaultVertexParallelismAndInputInfosDecider
+        implements VertexParallelismAndInputInfosDecider {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(DefaultVertexParallelismAndInputInfosDecider.class);
+
+    /**
+     * The cap ratio of broadcast bytes to data volume per task. The cap ratio is 0.5 currently
+     * because we usually expect the broadcast dataset to be smaller than non-broadcast. We can make
+     * it configurable later if we see users requesting for it.
+     */
+    private static final double CAP_RATIO_OF_BROADCAST = 0.5;
+
+    private final int maxParallelism;
+    private final int minParallelism;
+    private final long dataVolumePerTask;
+    private final int defaultSourceParallelism;
+
+    private DefaultVertexParallelismAndInputInfosDecider(
+            int maxParallelism,
+            int minParallelism,
+            MemorySize dataVolumePerTask,
+            int defaultSourceParallelism) {
+
+        checkArgument(minParallelism > 0, "The minimum parallelism must be larger than 0.");
+        checkArgument(
+                maxParallelism >= minParallelism,
+                "Maximum parallelism should be greater than or equal to the minimum parallelism.");
+        checkArgument(
+                defaultSourceParallelism > 0,
+                "The default source parallelism must be larger than 0.");
+        checkNotNull(dataVolumePerTask);
+
+        this.maxParallelism = maxParallelism;
+        this.minParallelism = minParallelism;
+        this.dataVolumePerTask = dataVolumePerTask.getBytes();
+        this.defaultSourceParallelism = defaultSourceParallelism;
+    }
+
+    @Override
+    public ParallelismAndInputInfos decideParallelismAndInputInfosForVertex(
+            List<BlockingResultInfo> consumedResults, int initialParallelism) {
+        checkArgument(
+                initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                        || initialParallelism > 0);
+
+        if (initialParallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                && areAllInputsAllToAll(consumedResults)
+                && !areAllInputsBroadcast(consumedResults)) {
```

Comment:
I know this can work, but is a bit weird that we rely on this: for a source vertex, `areAllInputsBroadcast == true`.
I prefer to explicitly check if it is a source and goes to the right branch.